server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -977,31 +974,35 @@ void createRetentionLease(final long startingSeqNo, ActionListener<RetentionLeas
                 assert shard.indexSettings().getIndexVersionCreated().before(Version.V_7_4_0)
                     || shard.indexSettings().isSoftDeleteEnabled() == false;
                 final long estimatedGlobalCheckpoint = startingSeqNo - 1;
-                final var newLease = shard.addPeerRecoveryRetentionLease(targetNodeId, estimatedGlobalCheckpoint, delayedListener);
+                final var newLease = shard.addPeerRecoveryRetentionLease(targetNodeId, estimatedGlobalCheckpoint, leasesSyncedListener);
                 logger.trace("created retention lease with estimated checkpoint of [{}]", estimatedGlobalCheckpoint);
                 return newLease;
             }
-        }), shard, cancellableThreads, leaseListener.delegateResponse((l, e) -> outerListener.onFailure(e)));
+        }), shard, cancellableThreads, leaseCreatedStep);
+
+        // NB we release the operation permit as soon as we have created the lease, but delay the outer listener until it is synced
+        leaseCreatedStep.whenComplete(lease -> leasesSyncedStep.addListener(outerListener.map(ignored -> lease)), outerListener::onFailure);
     }
 
     private void deleteRetentionLease(ActionListener<Void> outerListener) {
-        // NB we release the operation permit as soon as we have deleted the lease, but delay the outer listener until it is synced
-        final var leaseListener = new SubscribableListener<Void>();
-        final var delayedListener = new ThreadedActionListener<>(
-            shard.getThreadPool().generic(),
-            outerListener.<Void>delegateFailure((l, ignored) -> leaseListener.addListener(l))
-        );
+        final var leaseDeletedStep = new StepListener<Void>();
+        final var leasesSyncedStep = new StepListener<Void>();
+        final var leasesSyncedListener = new ThreadedActionListener<>(shard.getThreadPool().generic(), leasesSyncedStep).<
+            ReplicationResponse>map(ignored -> null);
 
         runUnderPrimaryPermit(permitListener -> ActionListener.completeWith(permitListener, () -> {
             try {
-                shard.removePeerRecoveryRetentionLease(request.targetNode().getId(), delayedListener.map(ignored -> null));
+                shard.removePeerRecoveryRetentionLease(request.targetNode().getId(), leasesSyncedListener);
```

Comment:
Why not just inline the forking listener as:

```java
                shard.removePeerRecoveryRetentionLease(
                    request.targetNode().getId(),
                    new ThreadedActionListener<>(
                        shard.getThreadPool().generic(),
                        outerListener.delegateFailure((l, ignored) -> leaseListener.addListener(l))
                    )
                );
)
```

To me that makes everything a lot clearer. Not so sure about adding another listener tbh.

Reply:
I considered this but decided I preferred it if the two methods had the same overall shape. And I found it helped me to give this listener a name, to remind the reader why there's a listener involved here. At least for the create method, because it's weird that `cloneLocalPeerRecoveryRetentionLease` returns a value _and_ takes a listener.

I'd be happy to create this listener within the `runUnderPrimaryPermit` call tho. WDYT?

Reply:
I agree with the same structure comment. I wonder if maybe it would be easiest to understand this by adding a method:

```java
    /**
     * Creates a listener that executes the given {@code listener}'s callbacks on the generic thread pool. The resulting listener will 
     * forward exceptions directly to {@code listener} but will delay forwarding successful responses until {@code dependency} completes. 
     * @param listener listener to delay
     * @param dependency listener that must complete successfully before {@code listener} is resolved
     * @return delayed listener
     */
    private <T, S> ActionListener<T> delayedListener(ActionListener<S> listener, SubscribableListener<S> dependency) {
        return new ThreadedActionListener<>(
            shard.getThreadPool().generic(),
            listener.delegateFailure((l, ignored) -> dependency.addListener(l))
        );
    }
```

to create these listeners that has some javadoc that clearly explains what the rules are for the delay (you probably can do better than I on the "clearly" here). I was a little hard for me to work out how the delay comes into things exactly in this code and I wrote a lot of the listener magic used :D, might be easier to just put things in words rather than add even more listeners? :)

Reply:
Yeah ok that's a fair point, we can make the common structure explicit. And add comments. See 6fc7d5a0192.